src/main/java/net/snowflake/ingest/connection/OAuthManager.java

```diff
@@ -0,0 +1,175 @@
+
+  // Properties for token refresh request
+  private static final String TOKEN_TYPE = "OAUTH";
+
+  // Update threshold, a floating-point value representing the ratio between the expiration time of
+  // an access token and the time needed to update it. It must be a value greater than 0 and less
+  // than 1. E.g. An access token with expires_in=600 and update_threshold_ratio=0.8 would be
+  // updated after 600*0.8 = 480.
+  private final double updateThresholdRatio;
+
+  private OAuthClient oAuthClient;
+
+  /**
+   * Creates a OAuthManager entity for a given account, user and OAuthCredential with default time
+   * to refresh the access token
+   *
+   * @param accountName - the snowflake account name of this user
+   * @param username - the snowflake username of the current user
+   * @param oAuthCredential - the OAuth credential we're using to connect
+   * @param baseURIBuilder - the uri builder with common scheme, host and port
+   * @param telemetryService reference to the telemetry service
+   */
+  OAuthManager(
+      String accountName,
+      String username,
+      OAuthCredential oAuthCredential,
+      URIBuilder baseURIBuilder,
+      TelemetryService telemetryService) {
+    this(
+        accountName,
+        username,
+        oAuthCredential,
+        baseURIBuilder,
+        DEFAULT_UPDATE_THRESHOLD_RATIO,
+        telemetryService);
+  }
+
+  /**
+   * Creates a OAuthManager entity for a given account, user and OAuthCredential with a specified
+   * time to refresh the token
+   *
+   * @param accountName - the snowflake account name of this user
+   * @param username - the snowflake username of the current user
+   * @param oAuthCredential - the OAuth credential we're using to connect
+   * @param baseURIBuilder - the uri builder with common scheme, host and port
+   * @param updateThresholdRatio - the ratio between the expiration time of a token and the time
+   *     needed to refresh it.
+   * @param telemetryService reference to the telemetry service
+   */
+  OAuthManager(
+      String accountName,
+      String username,
+      OAuthCredential oAuthCredential,
+      URIBuilder baseURIBuilder,
+      double updateThresholdRatio,
+      TelemetryService telemetryService) {
+    // disable telemetry service until jdbc v3.13.34 is released
+    super(accountName, username, null);
+
+    // if any of our arguments are null, throw an exception
+    if (oAuthCredential == null || baseURIBuilder == null) {
+      throw new IllegalArgumentException();
+    }
+
+    // check if update threshold is within (0, 1)
+    if (updateThresholdRatio <= 0 || updateThresholdRatio >= 1) {
+      throw new IllegalArgumentException("updateThresholdRatio should fall in (0, 1)");
+    }
+    this.updateThresholdRatio = updateThresholdRatio;
+    this.oAuthClient = new SnowflakeOAuthClient(accountName, oAuthCredential, baseURIBuilder);
+
+    // generate our first token
+    refreshToken();
+  }
```

Comment:
Not covered by tests?

Reply:
For this #L94-L95 and #L50-L57, those lines require actual snowflake oauth client to run. These lines required backdoor function which cannot be test in this repo, the tests are added in the dew test [PR](https://github.com/snowflakedb/snowflake/pull/110184).